daemon/api_download.go

```diff
@@ -68,38 +108,154 @@ func postSnapDownload(c *Command, r *http.Request, user *auth.UserState) Respons
 }
 
 func streamOneSnap(c *Command, action snapDownloadAction, user *auth.UserState) Response {
-	actions := []*store.SnapAction{{
-		Action:       "download",
-		InstanceName: action.SnapName,
-		Revision:     action.Revision,
-		CohortKey:    action.CohortKey,
-		Channel:      action.Channel,
-	}}
-	sars, err := getStore(c).SnapAction(context.TODO(), nil, actions, user, nil)
+	secret, err := downloadTokensSecret(c)
 	if err != nil {
-		return errToResponse(err, []string{action.SnapName}, InternalError, "cannot download snap: %v")
+		return InternalError(err.Error())
+	}
+	theStore := getStore(c)
+
+	var ss *snapStream
+	if action.ResumeToken == "" {
+		var info *snap.Info
+		actions := []*store.SnapAction{{
+			Action:       "download",
+			InstanceName: action.SnapName,
+			Revision:     action.Revision,
+			CohortKey:    action.CohortKey,
+			Channel:      action.Channel,
+		}}
+		sars, err := theStore.SnapAction(context.TODO(), nil, actions, user, nil)
```

Comment:
I think the name `results` would be more readable than `sars`, not taking into account the negative association of that acronym.